pytket/extensions/cutensornet/tnstate/general.py

```diff
@@ -0,0 +1,422 @@
+
+    @abstractmethod
+    def get_statevector(self) -> np.ndarray:
+        """Returns the statevector with qubits in Increasing Lexicographic Order (ILO).
+
+        Raises:
+            ValueError: If there are no qubits left in the state.
+        """
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def get_amplitude(self, state: int) -> complex:
+        """Returns the amplitude of the chosen computational state.
+
+        Notes:
+            The result is equivalent to ``tnstate.get_statevector[b]``, but this method
+            is faster when querying a single amplitude (or just a few).
+
+        Args:
+            state: The integer whose bitstring describes the computational state.
+                The qubits in the bitstring are in increasing lexicographic order.
+
+        Returns:
+            The amplitude of the computational state in ``self``.
+        """
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def get_qubits(self) -> set[Qubit]:
+        """Returns the set of qubits that ``self`` is defined on."""
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def get_byte_size(self) -> int:
+        """Returns the number of bytes ``self`` currently occupies in GPU memory."""
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def get_device_id(self) -> int:
+        """Returns the identifier of the device (GPU) where the tensors are stored."""
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def update_libhandle(self, libhandle: CuTensorNetHandle) -> None:
+        """Update the ``CuTensorNetHandle`` used by ``self``. Multiple
+        objects may use the same handle.
+
+        Args:
+            libhandle: The new cuTensorNet library handle.
+
+        Raises:
+            RuntimeError: If the device (GPU) where ``libhandle`` was initialised
+                does not match the one where the tensors of ``self`` are stored.
+        """
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def copy(self) -> TNState:
+        """Returns a deep copy of ``self`` on the same device."""
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+    @abstractmethod
+    def _flush(self) -> None:
+        raise NotImplementedError(f"Method not implemented in {type(self).__name__}.")
+
+
+def _safe_qr(
+    libhandle: CuTensorNetHandle, indices: str, T: cp.ndarray
+) -> tuple[cp.ndarray, cp.ndarray]:
+    """Wrapper of cuTensorNet QR decomposition to work around a bug in 23.6.0.
+
+    The bug has been reported https://github.com/NVIDIA/cuQuantum/discussions/96.
+    """

```

Comment:
Latest cuTensorNet version fixed this bug, so this function is no longer needed.

Reply:
Done